src/server/services/wfs/qgswfsutils.cpp

```diff
@@ -153,6 +153,47 @@ namespace QgsWfs
       request.setFlags( QgsFeatureRequest::ExactIntersect | QgsFeatureRequest::NoFlags );
       return request;
     }
+    // Apply BBOX through filterRect even inside an And to use spatial index
+    else if ( filterElem.firstChildElement().tagName() == QLatin1String( "And" ) &&
+              !filterElem.firstChildElement().firstChildElement( QLatin1String( "BBOX" ) ).isNull() )
+    {
+      QDomElement childElem = filterElem.firstChildElement().firstChildElement();
+      while ( !childElem.isNull() )
+      {
+        QDomElement childFilterElement = filterElem.ownerDocument().createElement( QLatin1String( "Filter" ) );
+        childFilterElement.appendChild( childElem.cloneNode( true ) );
+        QgsFeatureRequest childRequest = parseFilterElement( typeName, childFilterElement );
+        if ( childElem.tagName() == QLatin1String( "BBOX" ) )
+        {
+          if ( request.filterRect().isEmpty() )
+          {
+            request.setFilterRect( childRequest.filterRect() );
+          }
+          else
+          {
+            request.setFilterRect( request.filterRect().intersect( &childRequest.filterRect() ) );
+          }
+        }
+        else
+        {
+          if ( !bool( request.filterExpression() ) )
+          {
+            request.setFilterExpression( childRequest.filterExpression()->expression() );
+          }
+          else
+          {
+            QgsExpressionNode *opLeft = request.filterExpression()->rootNode()->clone();
+            QgsExpressionNode *opRight = childRequest.filterExpression()->rootNode()->clone();
+            QgsExpressionNode *node = new QgsExpressionNodeBinaryOperator( QgsExpressionNodeBinaryOperator::boAnd, opLeft, opRight );
```

Comment:
memory leak?

Reply:
I'm not completely aware with memory leaks but, after some debugging, it seems that deleting `node` also delete all child nodes, and other expressions and related nodes are also deleted with QgsFeatureRequest object. So it should be ok now with `delete node`.

Reply:
Yep, looks better now.
It would be prefect if you could use
`std::unique_ptr<QgsExpressionNodeBinaryOperator> node = std::make_unique<QgsExpressionNodeBinaryOperator>( ... );`

Reply:
Thanks for the update.
Did the `make_unique` not work?

Reply:
I missed the make_unique, done with qgis::make_unique.